signac/core/synceddict.py

```diff
@@ -148,20 +105,32 @@ def _save(self):
         pass
 
     def _dfs_update(self, old, new):
-        for key in new:
-            if key in old:
-                if old[key] == new[key]:
-                    continue
-                elif isinstance(new[key], Mapping) and isinstance(old[key], Mapping):
-                    self._dfs_update(old[key], new[key])
-                    continue
-            old[key] = self._dfs_convert(new[key])
-        remove = set()
-        for key in old:
-            if key not in new:
-                remove.add(key)
-        for key in remove:
-            del old[key]
+        if isinstance(old, Mapping) and isinstance(new, Mapping):
+            for key in new:
+                if key in old:
+                    if old[key] == new[key]:
+                        continue
+                    elif isinstance(new[key], Collection) and isinstance(old[key], Collection):
+                        self._dfs_update(old[key], new[key])
+                        continue
+                old[key] = self._dfs_convert(new[key])
+            remove = set()
+            for key in old:
+                if key not in new:
+                    remove.add(key)
+            for key in remove:
+                del old[key]
+        elif _is_nonstring_sequence(old) and _is_nonstring_sequence(new):
+            for index, item in enumerate(new):
+                if index < len(old):
+                    if old[index] == new[index]:
+                        continue
+                    elif isinstance(new[index], Collection) and isinstance(old[index], Collection):
+                        self._dfs_update(old[index], new[index])
+                        continue
+                old[index] = self._dfs_convert(new[index])
+            for index in range(len(new), len(old)):
+                del old[index]
```

Comment:
I tried that earlier and it failed because it was calling `_dfs_update` recursively on strings (which are treated as an instance of `Collection`). I attempted to stop the recursion if `old[key]` or `new[key]` are instances of `six.string_types` but that seems to cause data not to be updated.